src/main/java/com/lothrazar/cyclic/block/endershelf/EnderShelfItemHandler.java

```diff
@@ -0,0 +1,246 @@
+        });
+      }
+      else
+        return stack;
+    }
+    //if (isFakeSlot(slot) && !remaining.get().isEmpty())
+      //remaining.set(super.insertItem(slot, remaining.get(), simulate));
+    if (!isFakeSlot(slot))
+      remaining.set(super.insertItem(slot, stack, simulate));
+    if (!this.shelf.getWorld().isRemote && !simulate) {
+      PacketRegistry.sendToAllClients(this.shelf.getWorld(), new PacketTileInventory(this.shelf.getPos(), slot, this.getStackInSlot(slot), PacketTileInventory.TYPE.SET));
+    }
+    return remaining.get();
+  }
+
+  private boolean isFakeSlot(int slot) {
+    return slot == getFakeSlot();
+  }
+
+  private int getFakeSlot() {
+    return this.getSlots() - 1;
+  }
+
+  public BlockPos findShelfForInsert(ItemStack stack) {
+    //Set<BlockPos> connectedShelves = findConnectedShelves(new HashSet<>(), this.shelf.getWorld(), this.shelf.getPos(), 0);
+    //Set<BlockPos> connectedShelves = findConnectedShelves(this.shelf.getWorld(), this.shelf.getPos());
+    System.out.println("+++++ STARTING SEARCH ++++++");
+    //Set<BlockPos> connectedShelves = findConnectedShelves(new HashMap<BlockPos, Integer>(), this.shelf.getWorld(), this.shelf.getPos(), 0).keySet();
+    Set<BlockPos> connectedShelves = findConnectedShelves(this.shelf.getWorld(), this.shelf.getPos());
+    System.out.println("+++++ ENDING SEARCH ++++++");
+    Set<BlockPos> shelvesWithFreeSlots = new HashSet<>();
+    BlockPos originalPos = this.shelf.getPos();
+
+    Optional<BlockPos> possibleOtherDestination = connectedShelves.stream().filter(pos -> {
+      TileEntity otherTile = this.shelf.getWorld().getTileEntity(pos);
+      if (otherTile instanceof TileEnderShelf) {
+        if (otherTile.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY).isPresent()) {
+          EnderShelfItemHandler handler = (EnderShelfItemHandler) otherTile.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY).resolve().get();
+          for (int i = 0; i < handler.getSlots() - 1; i++) {
+            ItemStack otherStack = handler.getStackInSlot(i);
+            //System.out.println("Other stack is " + otherStack.toString());
+            if (otherStack.isEmpty()) {
+              if (!shelvesWithFreeSlots.contains(pos))
+                //System.out.println("Added " + pos.toString() + " to free slot list");
+              shelvesWithFreeSlots.add(pos);
+            }
+            if (pos != originalPos && UtilEnchant.doBookEnchantmentsMatch(stack, otherStack) && otherStack.getCount() < handler.getStackLimit(i, stack)) {
+              //System.out.println("Candidate at " + pos.toString());
+              return true;
+            }
+            //System.out.println("Ruled out " + pos.toString());
+          }
+        }
+      }
+      return false;
+    }).findFirst();
+    BlockPos finals = possibleOtherDestination.orElse(shelvesWithFreeSlots.stream().findFirst().orElse(originalPos));
+    //System.out.println("Going to try " + finals.toString());
```

Comment:
I havent ran the code, just looking over thinking about the insert dupe when its full  

If all shelves are full and theres no spot in the book, should this return orElse(null)?  
If it returns originalPos== this.shelf.getPos() , then above  it wont hit the "else return stack" statement.

Reply:
From what I understand, orElse isn't even supposed to take null as an argument (it's @Nonnull).

But the section of code you commented on is where we're trying to find a valid shelf to put it in, it's not actually doing any insertion here.

The `findShelfForInsert` method does the following:
Searches recursively in a defined radius for connected shelves
Iterate over their item handlers doing two things -- adding shelves with free slots to the `shelvesWithFreeSlots` Set and adding shelves that already have the book and room to put it to a stream.
At the end of that stream, `findFirst()` gets the first shelf that already has the enchantment and room for one more and stores it as an `Optional<BlockPos>`.
Finally, either return that single matching position of a shelf, or the first one with a free slot, or the original position if both of those fail.